src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,249 @@
+
+  // The object preview may be undefined until the user expands the object
+  // in the scopes panel. This applies primarily to prototype objects.
+  if (preview) {
+    const objectProperties = preview.properties.map(b => b.name);
+    properties.unshift(...objectProperties);
+  } else {
+    if (["Object", "Array"].includes(object.className)) {
+      const prototypeProperties = Object.getOwnPropertyNames(Object.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+
+    if (object.className === "Array") {
+      const prototypeProperties = Object.getOwnPropertyNames(Array.prototype);
+      properties.unshift(...prototypeProperties);
+    }
+  }
+
+  const prototype = preview?.prototypeValue;
+
+  // Recursively gather the properties through the prototype chain.
+  if (prototype?._object) {
+    const prototypeProperties = getPropertiesForObject(prototype._object);
+    properties.unshift(...prototypeProperties);
+  }
+
+  return properties;
+}
+
+export function getParentExpression(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(0, input.lastIndexOf("["));
+  } else {
+    return input.slice(0, input.lastIndexOf("."));
+  }
+}
+
+export function shouldPropertyAutocomplete(input: string) {
+  if (!isAccessingObjProperty(input)) {
+    return false;
+  }
+
+  const parentExpression = getParentExpression(input);
+
+  if (!parentExpression) {
+    return false;
+  }
+
+  // Right now, this only works for direct property access (e.g., a.b)
+  // and doesn't support chained expressions (e.g., a.b.c.d). This is because
+  // we don't (yet) have the chained values in the fetched frame scope data.
+  return isIdentifier(parentExpression);
+}
+function shouldVariableAutocomplete(input: string) {
+  return isIdentifier(input);
+}
+
+function getBinding(name: string, frameScope: FrameScope) {
+  return frameScope.scope.bindings.find(b => b.name === name);
+}
+function getBindingNames(scope: FrameScope): string[] {
+  if (!scope?.scope) {
+    return [];
+  }
+
+  return scope.scope.bindings.map(b => b.name);
+}
+function getGlobalVariables(scopes: Scope) {
+  const variableNames = [];
+  const globalObject = scopes.parent.object;
+
+  if (globalObject) {
+    const properties = getPropertiesForObject(globalObject._object);
+    variableNames.push(...properties);
+  }
+
+  return variableNames;
+}
+
+export function getLastToken(input: string) {
+  if (isBracketNotation(input)) {
+    return input.slice(input.lastIndexOf("[") + 1);
+  } else {
+    return input.slice(input.lastIndexOf(".") + 1);
+  }
+}
+
+export function getCursorIndex(value: string) {
+  if (shouldPropertyAutocomplete(value)) {
+    // +1 to account for the `.` or `[`
+    return getParentExpression(value).length + 1;
+  } else {
+    return 0;
+  }
+}
+
+export function insertAutocompleteMatch(value: string, match: string) {
+  const parentExpression = getParentExpression(value);
+
+  if (shouldPropertyAutocomplete(value)) {
```

Comment:
Oh, I was reading this as `if this returns false we don't autocomplete`, but actually it's `if this returns false we variable autocomplete`?

Reply:
Exactly. The variable autocomplete was implicit in the else so I added a comment. How would you do it?

Reply:
It's probably a little late for this refactoring, but I think I would have three *types* of auto-complete: 

```
enum LineCompletion {
  DotProperty,
  BracketProperty,
  Variable
}
```

And then have each one handle its own matching/generating a new line. The idea here is that we might not know if this is *every* type of autocomplete or not. For instance, what if we added some built-in commands that didn't quite follow this pattern, now we have another set of nested `if/else`s to add to a bunch of places in this file. If we reframe the problem into "there are at least three types of auto-complete, and we're not 100% certain that there will never be any more, but we *are* certain that every auto-completion line will only ever be *one* type)" then this code gets easier to extend later. If you want to pair on making this change lmk, but also, if you would rather move on with life I will understand 😄